staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go

```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
 func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]string, error) {
 	zonedNodes := splitNodesByZone(nodes)
 	klog.V(2).Infof("ensureInternalInstanceGroups(%v): %d nodes over %d zones in region %v", name, len(nodes), len(zonedNodes), g.region)
+
 	var igLinks []string
-	for zone, nodes := range zonedNodes {
-		igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+	gceZonedNodes := map[string][]string{}
```

Comment:
As i understand this change, here is what happens:

1) For each of the zones that the cluster has nodes in, list all existing instance groups.
2) if any of the instance groups contain just cluster nodes(all of them or a subset) in the zone, use that instance group. Create a new instance group with just the remaining nodes.

I see this will help in case there are control-plane nodes that are in their own instance groups. How is it helping the other use-case, where nodes are in multiple subnets?

If the use-case is just to support the control-plane scenario, can we not label those nodes with a master role? https://github.com/kubernetes/kubernetes/blob/ed5b03831385b0071b127ff6c329ae344b9780f6/pkg/controller/service/controller.go#L626

Reply:
> So the subnet limitation should be automatically handled by the k8s cloud provider, but for now allowing users to create the IGs for instances that require this special setup should definietly help, and k8s cloud provider can just use those as-is, while maintaining the membership and lifecycle for ones created by it.

As from the PR description, k8s code doesn't handle nodes in different subnets, this allows users to create their IGs and k8s will reuse them until it can handle the subnets itself.

Reply:
Echoing the question above: is it not possible to use "master role" to exclude the master node from load balancing?

Reply:
> Echoing the question above: is it not possible to use "master role" to exclude the master node from load balancing?

It is most definitely possible, that only applies when the nodes are master nodes, that wouldn't solve the problem when the node is actually a compute node.. ?

Reply:
Also we want our master nodes to be part of the LB backend, there is no reason why the control-plane hosts cannot be part of the LB backend pool.

Reply:
> > Echoing the question above: is it not possible to use "master role" to exclude the master node from load balancing?
> 
> It is most definitely possible, that only applies when the nodes are master nodes, that wouldn't solve the problem when the node is actually a compute node.. ?

Is there a use-case for this? As I understood the problem, the master nodes that were part of a different LoadBalancer Instance group.

Reply:
hmm...

the first case that we want to support is:
**We want masters to be part of the SLB backend.** 
But we have other services on the control-plane that cannot be backed by SLB yet, therefore we have separate externally managed LB pointing to the control-plane nodes.. and that blocks our masters need to be part of SLB as gcp only allows an instance to be part of only one instance-group that is backed by LB

reusing the instance-groups for SLB allows us to achieve external LB pointing to control-plane and SLBs pointing to control.

The second case that we want to support is:
**nodes that are backends of the SLB can be in mutiple subnets,** 
today that is not possible as the k8s controller creates and tries to put all the instances in a zone in one instance-group irrespective of the subnet.

Reusing the instance-groups allows us to manually create the instance-groups for now and let k8s create and manage the rest in the same subnet.

Reply:
I'll note that this is unique to GCP - other cloud providers work fine when masters have SLB or external cloud LB pointed at them.  So the first part abhinav mentioned is a consistency and behavior (we wouldn't want GCP to be inconsistent with the others, etc).

On the subnets thing, there are lots of reasons for instances in a zone to be in different subnets (you may want a set of DMZ nodes targeted by internal traffic that have firewall rules to block all but a specific set of config from crossing).